

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>velocyto.counter &mdash; velocyto 0.13.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="velocyto 0.13.1 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> velocyto
          

          
          </a>

          
            
            
              <div class="version">
                0.13
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to velocyto.py!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../fullapi/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/index.html">Notebooks</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../citing/index.html">Authors and citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">velocyto</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>velocyto.counter</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for velocyto.counter</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="k">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">chain</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">velocyto</span> <span class="k">as</span> <span class="nn">vcy</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>


<div class="viewcode-block" id="ExInCounter"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter">[docs]</a><span class="k">class</span> <span class="nc">ExInCounter</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot; Main class to do the counting of introns and exons &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">logic</span><span class="p">:</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Logic</span><span class="p">,</span> <span class="n">valid_bcset</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; valid_bcs2idx is a dict mapping valid cell barcodes to unique arbitrary numeric indexes used in the output arrays &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logic</span> <span class="o">=</span> <span class="n">logic</span><span class="p">()</span>
        <span class="c1"># NOTE: maybe there shoulb be a self.logic.validate() step at the end of init</span>
        <span class="k">if</span> <span class="n">valid_bcset</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcset</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcset</span> <span class="o">=</span> <span class="n">valid_bcset</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>  <span class="c1"># type: Dict[str, List]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">GeneInfo</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># NOTE: by using a default dict and not logging access to keys that do not exist, we might miss bugs!!!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">test_flag</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">bclen</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_bcset</span><span class="p">)))</span>
        <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
    
<div class="viewcode-block" id="ExInCounter.parse_cigar_tuple"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.parse_cigar_tuple">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_cigar_tuple</span><span class="p">(</span><span class="n">cigartuples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">],</span> <span class="n">pos</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
        <span class="n">segments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">hole_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">ref_skip</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">clip5</span> <span class="o">=</span> <span class="n">clip3</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="k">for</span> <span class="n">operation_id</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">cigartuples</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># vcy.CIGAR[operation_id] == &quot;BAM_CMATCH&quot;</span>
                <span class="n">segments</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">p</span><span class="p">,</span> <span class="n">p</span> <span class="o">+</span> <span class="n">length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>  <span class="c1"># A splice || vcy.CIGAR[operation_id] == &#39;BAM_CREF_SKIP&#39;</span>
                <span class="n">ref_skip</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>  <span class="c1"># A deletion || cy.CIGAR[operation_id] == &#39;BAM_CDEL&#39;</span>
                <span class="k">if</span> <span class="n">length</span> <span class="o">&lt;=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">PATCH_INDELS</span><span class="p">:</span>
                    <span class="n">hole_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">segments</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># bases at 5&#39; or 3&#39; are NOT part of the alignment || vcy.CIGAR[operation_id] == &#39;BAM_CSOFT_CLIP&#39;</span>
                <span class="k">if</span> <span class="n">p</span> <span class="o">==</span> <span class="n">pos</span><span class="p">:</span>
                    <span class="n">clip5</span> <span class="o">=</span> <span class="n">length</span>  <span class="c1"># At start of alignment</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">clip3</span> <span class="o">=</span> <span class="n">length</span>  <span class="c1"># Must be at end of alignment vcy.CIGAR[operation_id] in [&quot;BAM_CINS&quot;, &quot;BAM_CHARD_CLIP&quot;]</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># An insertion BAM_CINS</span>
                <span class="k">if</span> <span class="n">length</span> <span class="o">&lt;=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">PATCH_INDELS</span><span class="p">:</span>
                    <span class="n">hole_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">segments</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
                <span class="c1"># else do nothing</span>
                <span class="c1"># NOTE: maybe we should make so that the reads get discarded</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>  <span class="c1"># BAM_CHARD_CLIP</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Hard clip was encountered! All mapping are assumed soft clipped&quot;</span><span class="p">)</span>

        <span class="c1"># Merge segments separated by small insertions and deletions</span>
        <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">hole_to_remove</span><span class="p">):</span>
            <span class="n">segments</span><span class="p">[</span><span class="n">b</span> <span class="o">-</span> <span class="n">a</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">segments</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">b</span> <span class="o">-</span> <span class="n">a</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="n">segments</span><span class="p">[</span><span class="n">b</span> <span class="o">-</span> <span class="n">a</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
</div>
        <span class="k">return</span> <span class="n">segments</span><span class="p">,</span> <span class="n">ref_skip</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span>

<div class="viewcode-block" id="ExInCounter.peek"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.peek">[docs]</a>    <span class="k">def</span> <span class="nf">peek</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">lines</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Peeks into the samfile to determine if it is a cellranger or dropseq file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Peeking into </span><span class="si">{samfile}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">fin</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">samfile</span><span class="p">)</span>  <span class="c1"># type: pysam.AlignmentFile</span>
        <span class="n">cellranger</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">dropseq</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">failed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">read</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fin</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s2">&quot;CB&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s2">&quot;UB&quot;</span><span class="p">):</span>
                <span class="n">cellranger</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s2">&quot;XC&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s2">&quot;XM&quot;</span><span class="p">):</span>
                <span class="n">dropseq</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Not found cell and umi barcode in entry </span><span class="si">{i}</span><span class="s2"> of the bam file&quot;</span><span class="p">)</span>
                <span class="n">failed</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">cellranger</span> <span class="o">&gt;</span> <span class="n">lines</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">cellbarcode_str</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">umibarcode_str</span> <span class="o">=</span> <span class="s2">&quot;UB&quot;</span>
                <span class="k">break</span>
            <span class="k">elif</span> <span class="n">dropseq</span> <span class="o">&gt;</span> <span class="n">lines</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">cellbarcode_str</span> <span class="o">=</span> <span class="s2">&quot;XC&quot;</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">umibarcode_str</span> <span class="o">=</span> <span class="s2">&quot;XM&quot;</span>
                <span class="k">break</span>
            <span class="k">elif</span> <span class="n">failed</span> <span class="o">&gt;</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">lines</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;The bam file does not contain cell and umi barcodes appropriatelly formatted&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">pass</span></div>
        <span class="n">fin</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<div class="viewcode-block" id="ExInCounter.iter_alignments"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.iter_alignments">[docs]</a>    <span class="k">def</span> <span class="nf">iter_alignments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">unique</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">yield_line</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Iterates over the bam/sam file and yield Read objects</span>

<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        samfile: str</span>
<span class="sd">            path to the sam file</span>
<span class="sd">        unique: bool</span>
<span class="sd">            yield only unique allignments</span>
<span class="sd">        yield_line: bool</span>
<span class="sd">            whether to yield the raw sam line</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        yields vcy.Read for each valid line of the bam file</span>
<span class="sd">        or a Tuple (vcy.Read, sam_line) if ``yield_line==True``</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># No peeking here, this will happen a layer above, and only once  on the not sorted file! before it was self.peek(samfile, lines=10)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Reading </span><span class="si">{samfile}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">fin</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">samfile</span><span class="p">)</span>  <span class="c1"># type: pysam.AlignmentFile</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">read</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fin</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10000000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Read first {i // 1000000} million reads&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="c1"># If unique parameter is set to True, skip not unique allignments</span>
            <span class="k">if</span> <span class="n">unique</span> <span class="ow">and</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s2">&quot;NH&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cellbarcode_str</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># NOTE: this rstrip is relevant only for cellranger, should not cause trouble in Dropseq</span>
                <span class="n">umi</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">umibarcode_str</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>  <span class="c1"># NOTE: Here errors could go unnoticed</span>
            <span class="k">if</span> <span class="n">bc</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcset</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcset</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bc</span><span class="p">)</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">else</span> <span class="s1">&#39;+&#39;</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">fin</span><span class="o">.</span><span class="n">get_reference_name</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">rname</span><span class="p">)</span>  <span class="c1"># this is return a string otherwise read.rname for the integer</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">):</span>
                <span class="c1"># I have to deal with incongruences with the .gft (what is cellranger doing???)</span>
                <span class="c1"># NOTE Why this happens?</span>
                <span class="k">if</span> <span class="s2">&quot;_&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
                    <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;M&quot;</span><span class="p">:</span>
                        <span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;MT&quot;</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">+</span> <span class="mi">1</span>  <span class="c1"># reads in pysam are always 0-based, but 1-based is more convenient to wor with in bioinformatics</span>
            <span class="n">segments</span><span class="p">,</span> <span class="n">ref_skipped</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parse_cigar_tuple</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">cigartuples</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">segments</span> <span class="o">==</span> <span class="p">[]:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;No segments in read:</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">)</span>

            <span class="n">read_object</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">(</span><span class="n">bc</span><span class="p">,</span> <span class="n">umi</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">segments</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span><span class="p">,</span> <span class="n">ref_skipped</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">yield_line</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read_object</span><span class="o">.</span><span class="n">span</span> <span class="o">&gt;</span> <span class="mi">3000000</span><span class="p">:</span>  <span class="c1"># Longest locus existing</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Trashing read, too long span</span><span class="se">\n</span><span class="s2">{read.tostring(fin)}&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">read_object</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">tostring</span><span class="p">(</span><span class="n">fin</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read_object</span><span class="o">.</span><span class="n">span</span> <span class="o">&gt;</span> <span class="mi">3000000</span><span class="p">:</span>  <span class="c1"># Longest locus existing</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Trashing read, too long span</span><span class="se">\n</span><span class="s2">{read.tostring(fin)}&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">read_object</span></div>
        <span class="n">fin</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<div class="viewcode-block" id="ExInCounter.read_repeats"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.read_repeats">[docs]</a>    <span class="k">def</span> <span class="nf">read_repeats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">tolerance</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;Read repeats and merge close ones into highly repetitive areas</span>
<span class="sd">        </span>
<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        gtf_file: str</span>
<span class="sd">            file to read</span>
<span class="sd">        tolerance: int, default=5</span>
<span class="sd">            if two repeats intervals to be masked are found closer than tolerance bases from each other they are fused in one bigger masked interval.</span>
<span class="sd">            Notice that in the downstream analysis only reads that are fall inside mask intervals are discarded</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        mask_ivls_by_chromstrand: Dict[str, List[vcy.Feature]]</span>
<span class="sd">            A dictionary key: chromosome+strand value: list of features (repeat intervals)</span>
<span class="sd">            (The reference is returned but an internal attribure self.self.masked_by_chrm_strand is kept)</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Example code to sort the gtf file</span>
        <span class="c1"># sorted_filename = gtffile.split(&quot;.&quot;)[0] + &quot;_sorted.gtf&quot;</span>
        <span class="c1"># logging.debug(f&quot;Sorting by `sort -k1,1 -k7,7 -k4,4n {gtffile} &gt; {sorted_filename}`&quot;)</span>
        <span class="c1"># with open(sorted_filename, &quot;w&quot;) as f:</span>
        <span class="c1">#     p1 = subprocess.run([&quot;sort&quot;, &quot;-k1,1&quot;, &quot;-k7,7&quot;, &quot;-k4,4n&quot;, gtffile],</span>
        <span class="c1">#                         stdout=f)</span>
            
        <span class="c1"># Parse arguments</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Reading </span><span class="si">{gtf_file}</span><span class="s1">, the file is assumed sorted (if is not interupt the process and sort it by running: ``sort -k1,1 -k7,7 -k4,4n``)&#39;</span><span class="p">)</span>
        
        <span class="c1"># Read up skipping headers up to the first valid entry</span>
        <span class="n">repeat_ivls_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># fin = open(gtf_file)</span>
        <span class="n">gtf_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">gtf_file</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>

        <span class="k">def</span> <span class="nf">sorting_key</span><span class="p">(</span><span class="n">entry</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
            <span class="sd">&quot;&quot;&quot;This sorting strategy is equivalent to sort -k1,1 -k7,7 -k4,4n&quot;&quot;&quot;</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="n">entry</span><span class="p">)</span>  <span class="c1"># The last element of the touple corresponds to the `last resort comparison`</span>
        
        <span class="n">gtf_lines</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gtf_lines</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sorting_key</span><span class="p">)</span>
        <span class="c1">###</span>

        <span class="n">line</span> <span class="o">=</span> <span class="n">gtf_lines</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">chrom</span><span class="p">,</span> <span class="n">feature_class</span><span class="p">,</span> <span class="n">feature_type</span><span class="p">,</span> <span class="n">start_str</span><span class="p">,</span> <span class="n">end_str</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">fields</span>
        <span class="c1"># Removing chr from the chromosome name to uniform different formats of gtf files, taht might or might not have the prefix &quot;chr&quot;</span>
        <span class="k">if</span> <span class="n">chrom</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;chr&quot;</span><span class="p">:</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start_str</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end_str</span><span class="p">)</span>
        <span class="n">chromstrand</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>

        <span class="c1"># Set this tu the current entry</span>
        <span class="n">curr_chrom</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">chrom</span>
        <span class="n">curr_feature_class</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">feature_class</span>
        <span class="n">curr_feature_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">feature_type</span>
        <span class="n">curr_start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">start</span>
        <span class="n">curr_end</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">end</span>
        <span class="n">curr_n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">curr_strand</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">strand</span>
        <span class="n">curr_tags</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">tags</span>
        <span class="n">curr_chromstrand</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">chromstrand</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gtf_lines</span><span class="p">:</span>
            
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">feature_class</span><span class="p">,</span> <span class="n">feature_type</span><span class="p">,</span> <span class="n">start_str</span><span class="p">,</span> <span class="n">end_str</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">fields</span>
            <span class="c1"># Removing chr from the chromosome name to uniform different formats of gtf files, taht might or might not have the prefix &quot;chr&quot;</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;chr&quot;</span><span class="p">:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start_str</span><span class="p">)</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end_str</span><span class="p">)</span>
            <span class="n">chromstrand</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>
            
            <span class="c1"># On chromosome or strand change: save and clean after yourself</span>
            <span class="k">if</span> <span class="n">chromstrand</span> <span class="o">!=</span> <span class="n">curr_chromstrand</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="p">[</span><span class="n">curr_chromstrand</span><span class="p">]</span> <span class="o">=</span> <span class="n">repeat_ivls_list</span>
                <span class="n">repeat_ivls_list</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">curr_chrom</span> <span class="o">=</span> <span class="n">chrom</span>
                <span class="n">curr_strand</span> <span class="o">=</span> <span class="n">strand</span>
                <span class="n">curr_chromstrand</span> <span class="o">=</span> <span class="n">curr_chrom</span> <span class="o">+</span> <span class="n">curr_strand</span>
                
            <span class="c1"># If there is no overlap/contiguity within a certain tolerance with the next</span>
            <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">curr_end</span> <span class="o">+</span> <span class="n">tolerance</span><span class="p">:</span>
                <span class="c1"># Close and store the previus interval</span>
                <span class="n">repeat_ivls_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">curr_start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">curr_end</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">),</span> <span class="n">exin_no</span><span class="o">=</span><span class="n">curr_n</span><span class="p">))</span>
                <span class="c1"># NOTE exin_no is being used to store curr_n, and kind could be used to store tag for some extra memory cost</span>
                
                <span class="c1"># Remember the newly parsed interval</span>
                <span class="n">curr_start</span> <span class="o">=</span> <span class="n">start</span>
                <span class="n">curr_end</span> <span class="o">=</span> <span class="n">end</span>
                <span class="n">curr_n</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># number of original repeat regions included</span>

                <span class="c1"># this is extra information that right now is not stored in vcy.Feature (not to waste memory) since is not used</span>
                <span class="n">curr_feature_class</span> <span class="o">=</span> <span class="n">feature_class</span>
                <span class="n">curr_feature_type</span> <span class="o">=</span> <span class="n">feature_type</span>
                <span class="n">curr_tags</span> <span class="o">=</span> <span class="n">tags</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Extend the masked interval!</span>
                <span class="n">curr_end</span> <span class="o">=</span> <span class="n">end</span>
                <span class="n">curr_n</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># this is extra information that right now is not saved (not to waste memory) since is not used</span>
                <span class="n">gap</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="n">curr_end</span>
                <span class="n">curr_tags</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{curr_tags}</span><span class="s2"> gap </span><span class="si">{gap}</span><span class="s2">; </span><span class="si">{tags}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">gap</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">curr_tags</span> <span class="o">+</span> <span class="n">tags</span>

        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chromstrand</span><span class="p">,</span> <span class="n">feature_list</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">feature_list</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># relies on the __lt__ method of vcy.Feature</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">feature_list</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Processed masked annotation .gtf and generated </span><span class="si">{n}</span><span class="s1"> intervals to mask!&#39;</span><span class="p">)</span>
</div>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span>

<div class="viewcode-block" id="ExInCounter.assign_indexes_to_genes"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.assign_indexes_to_genes">[docs]</a>    <span class="k">def</span> <span class="nf">assign_indexes_to_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">features</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Assign to each newly encoutered genes an unique indexes corresponding to the output matrix column ix</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Assigning indexes to genes&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">trmodel</span> <span class="ow">in</span> <span class="n">features</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">start</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">start</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span><span class="o">.</span><span class="n">end</span> <span class="o">&lt;</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">end</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">end</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">)</span></div>
                <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">GeneInfo</span><span class="p">(</span><span class="n">trmodel</span><span class="o">.</span><span class="n">genename</span><span class="p">,</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">geneid</span><span class="p">,</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">chromstrand</span><span class="p">,</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">trmodel</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>

<div class="viewcode-block" id="ExInCounter.read_transcriptmodels"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.read_transcriptmodels">[docs]</a>    <span class="k">def</span> <span class="nf">read_transcriptmodels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;Reads transcript models from a sorted .gtf file</span>

<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        gtf_file: str</span>
<span class="sd">            Path to the sorted gtf file</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        annotations_by_chrm_strand: Dict[str, List[vcy.TrancriptModel]]</span>
<span class="sd">            A dictionary key: chromosome+strand value: list of trascript models</span>
<span class="sd">            (The reference is returned but an internal attribure self.annotations_by_chrm_strand is kept)</span>

<span class="sd">        There will exist an object vcy.Features for the same exon appearing in a different vcy.TranscriptModel. (his is desired)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Define the regexes for the parsing</span>
        <span class="n">regex_trid</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;transcript_id &quot;([^&quot;]+)&quot;&#39;</span><span class="p">)</span>
        <span class="n">regex_trname</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;transcript_name &quot;([^&quot;]+)&quot;&#39;</span><span class="p">)</span>
        <span class="n">regex_geneid</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;gene_id &quot;([^&quot;]+)&quot;&#39;</span><span class="p">)</span>
        <span class="n">regex_genename</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;gene_name &quot;([^&quot;]+)&quot;&#39;</span><span class="p">)</span>
        <span class="n">regex_exonno</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;exon_number &quot;([^&quot;]+)&quot;&#39;</span><span class="p">)</span>

        <span class="c1"># Initialize containers</span>
        <span class="c1"># headerlines: List[str] = []</span>
        <span class="n">curr_chromstrand</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">features</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">]</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

        <span class="n">gtf_lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">gtf_file</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>

        <span class="k">def</span> <span class="nf">sorting_key</span><span class="p">(</span><span class="n">entry</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
            <span class="sd">&quot;&quot;&quot;This sorting strategy is equivalent to sort -k1,1 -k7,7 -k4,4n&quot;&quot;&quot;</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="n">entry</span><span class="p">)</span>  <span class="c1"># The last element of the touple corresponds to the `last resort comparison`</span>
        
        <span class="n">gtf_lines</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gtf_lines</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sorting_key</span><span class="p">)</span>
        <span class="c1"># Loop throug gtf file (assumes it is ordered)</span>
        <span class="k">for</span> <span class="n">nth_line</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gtf_lines</span><span class="p">):</span>
            <span class="c1"># Deal with headers</span>
            <span class="c1"># if line.startswith(&#39;#&#39;):</span>
            <span class="c1">#     headerlines.append(line)</span>
            <span class="c1">#     continue</span>
                
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">feature_class</span><span class="p">,</span> <span class="n">feature_type</span><span class="p">,</span> <span class="n">start_str</span><span class="p">,</span> <span class="n">end_str</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">junk</span><span class="p">,</span> <span class="n">tags</span> <span class="o">=</span> <span class="n">fields</span>

            <span class="c1"># Deal with possible incongruences between .gtf and bam file chromosome naming</span>
            <span class="c1"># We keep the name of the chromosome removing the chr from the name and removing part after the . if present</span>
            <span class="k">if</span> <span class="s2">&quot;chr&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">[:</span><span class="mi">4</span><span class="p">]:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>  <span class="c1"># NOTE before it was chrom[3:].split(&quot;.&quot;)[0]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">pass</span>  <span class="c1"># NOTE before it was chrom.split(&quot;.&quot;)[0]</span>

            <span class="c1"># A new chromosome/strand encountered</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span> <span class="o">!=</span> <span class="n">curr_chromstrand</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">curr_chromstrand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># Every time with exception with first and the last chromosome</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">:</span>
                        <span class="c1"># NOTE this is not enough as a check but it will detect with few checks if file is not sorted at all</span>
                        <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Genome annotation gtf file is not sorted correctly! Run the following command:</span><span class="se">\n</span><span class="s2">sort -k1,1 -k7,7 -k4,4n -o [GTF_OUTFILE] [GTF_INFILE]&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Done with </span><span class="si">{curr_chromstrand}</span><span class="s2"> [line {nth_line-1}]&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">assign_indexes_to_genes</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">curr_chromstrand</span><span class="p">]</span> <span class="o">=</span> <span class="n">features</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Seen {len(self.geneid2ix)} genes until now&quot;</span><span class="p">)</span>
                <span class="n">features</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Parsing Chromosome </span><span class="si">{chrom}</span><span class="s2"> strand </span><span class="si">{strand}</span><span class="s2"> [line </span><span class="si">{nth_line}</span><span class="s2">]&quot;</span><span class="p">)</span>
                <span class="n">curr_chromstrand</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>
                
            <span class="k">if</span> <span class="n">feature_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
                <span class="n">trid</span> <span class="o">=</span> <span class="n">regex_trid</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">tags</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">trname</span> <span class="o">=</span> <span class="n">regex_trname</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">tags</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">geneid</span> <span class="o">=</span> <span class="n">regex_geneid</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">tags</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">genename</span> <span class="o">=</span> <span class="n">regex_genename</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">tags</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">exonno</span> <span class="o">=</span> <span class="n">regex_exonno</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">tags</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start_str</span><span class="p">)</span>
                <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end_str</span><span class="p">)</span>
                <span class="n">chromstrand</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>
                
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">features</span><span class="p">[</span><span class="n">trid</span><span class="p">]</span><span class="o">.</span><span class="n">append_exon</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;e&quot;</span><span class="p">),</span> <span class="n">exin_no</span><span class="o">=</span><span class="n">exonno</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">features</span><span class="p">[</span><span class="n">trid</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">TranscriptModel</span><span class="p">(</span><span class="n">trid</span><span class="o">=</span><span class="n">trid</span><span class="p">,</span> <span class="n">trname</span><span class="o">=</span><span class="n">trname</span><span class="p">,</span> <span class="n">geneid</span><span class="o">=</span><span class="n">geneid</span><span class="p">,</span> <span class="n">genename</span><span class="o">=</span><span class="n">genename</span><span class="p">,</span> <span class="n">chromstrand</span><span class="o">=</span><span class="n">chromstrand</span><span class="p">)</span>
                    <span class="n">features</span><span class="p">[</span><span class="n">trid</span><span class="p">]</span><span class="o">.</span><span class="n">append_exon</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;e&quot;</span><span class="p">),</span> <span class="n">exin_no</span><span class="o">=</span><span class="n">exonno</span><span class="p">))</span>

        <span class="c1"># Do it for the last chromosome</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assign_indexes_to_genes</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">curr_chromstrand</span><span class="p">]</span> <span class="o">=</span> <span class="n">features</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Done with </span><span class="si">{curr_chromstrand}</span><span class="s2"> [line {nth_line-1}]&quot;</span><span class="p">)</span>
    
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Fixing corner cases of transcript models containg intron longer than {vcy.LONGEST_INTRON_ALLOWED//1000}Kbp&quot;</span><span class="p">)</span>
        <span class="c1"># Fix corner cases of extremelly long introns ~1Mbp that would be masking genes that are found internally</span>
        <span class="k">for</span> <span class="n">tmodels_orddict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">tm</span> <span class="ow">in</span> <span class="n">tmodels_orddict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">tm</span><span class="o">.</span><span class="n">chop_if_long_intron</span><span class="p">()</span>  <span class="c1"># Change it in place</span>

        <span class="c1"># Respect the sorting it had before</span>
        <span class="c1"># NOTE: not sure it is needed downstream anymore also not sure it guarantees exactly the same order</span>
        <span class="k">for</span> <span class="n">chromstrand</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">i</span><span class="o">.</span><span class="n">trid</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">chromstrand</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">chromstrand</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span>
</div>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span>

<div class="viewcode-block" id="ExInCounter.mark_up_introns"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.mark_up_introns">[docs]</a>    <span class="k">def</span> <span class="nf">mark_up_introns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">multimap</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Mark up introns that have reads across exon-intron junctions</span>
<span class="sd">        </span>
<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        samfile: str</span>
<span class="sd">            path to the bam or sam file to markup</span>
<span class="sd">        logic: vcy.Logic</span>
<span class="sd">            The logic object to use, changes in different techniques / levels of strictness</span>
<span class="sd">            NOTE: Right now it is not used</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Nothing it just add to validation to the vcy.Features</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        Situations not considered:</span>
<span class="sd">        # If an the exon is so short that is possible to get both exonA-exonB junction and exonB-intronB boundary in the same read</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># VERBOSE: dump_list = []</span>
        <span class="c1"># Read the file</span>
        <span class="n">currchrom</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignments</span><span class="p">(</span><span class="n">samfile</span><span class="p">,</span> <span class="n">unique</span><span class="o">=</span><span class="ow">not</span> <span class="n">multimap</span><span class="p">):</span>
            <span class="c1"># Don&#39;t consider spliced reads (exonic) in this step</span>
            <span class="c1"># NOTE Can the exon be so short that we get splicing and exon-intron boundary</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">is_spliced</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># When the chromosome mapping of the read changes, change interval index.</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span> <span class="o">!=</span> <span class="n">currchrom</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Marking up chromosome </span><span class="si">{r.chrom}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">currchrom</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span>
                <span class="k">if</span> <span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;The .bam file refers to a chromosome &#39;</span><span class="si">{currchrom}</span><span class="s2">+&#39; not present in the annotation (.gtf) file&quot;</span><span class="p">)</span>
                    <span class="n">sorted_features_f</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">sorted_features_f</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                <span class="k">if</span> <span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;The .bam file refers to a chromosome &#39;</span><span class="si">{currchrom}</span><span class="s2">-&#39; not present in the annotation (.gtf) file&quot;</span><span class="p">)</span>
                    <span class="n">sorted_features_r</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Feature</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">sorted_features_r</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
                <span class="n">iif</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">(</span><span class="n">sorted_features_f</span><span class="p">)</span>
                <span class="n">iir</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">(</span><span class="n">sorted_features_r</span><span class="p">)</span>
            
            <span class="c1"># Consider the correct strand</span>
            <span class="n">ii</span> <span class="o">=</span> <span class="n">iif</span> <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">iir</span>

            <span class="c1"># VERBOSE: # Look for overlap between the intervals and the read</span>
            <span class="c1"># VERBOSE: dump_list += ii.mark_overlapping_ivls(r)</span></div>
            <span class="n">ii</span><span class="o">.</span><span class="n">mark_overlapping_ivls</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

        <span class="c1"># VERBOSE: import pickle</span>
        <span class="c1"># VERBOSE: pickle.dump(dump_list, open(&quot;dump_mark_overlapping_ivls.pickle&quot;, &quot;wb&quot;))</span>

<div class="viewcode-block" id="ExInCounter.count"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.count">[docs]</a>    <span class="k">def</span> <span class="nf">count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">multimap</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">cell_batch_size</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">molecules_report</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot; Do the counting of molecules</span>
<span class="sd">        </span>
<span class="sd">        Arguments</span>
<span class="sd">        ---------</span>
<span class="sd">        samfile: str</span>
<span class="sd">            path to the bam or sam file to markup</span>
<span class="sd">        cell_batch_size: int, default = 50</span>
<span class="sd">            it defines whether to require or not exon-intron spanning read to consider an intron valid.</span>
<span class="sd">        molecules_report: bool, default=False</span>
<span class="sd">            whether to output a file that reports the intervals hit for each molecule counted</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        list_spliced_arrays, list_unspliced_arrays, list_ambiguous_arrays, cell_bcs_order</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        The memory footprint could be reduced allocating scipy sparse arrays</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize variables and containers</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># self.cells_since_last_count = 0</span>
        <span class="c1"># Analysis is cell wise so the Feature Index swapping is happening often and it is worth to preload everything in memory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">feature_indexes</span><span class="p">:</span> <span class="n">DefaultDict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chromstrand_key</span><span class="p">,</span> <span class="n">annotions_ordered_dict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">feature_indexes</span><span class="p">[</span><span class="n">chromstrand_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">annotions_ordered_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">mask_indexes</span><span class="p">:</span> <span class="n">DefaultDict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chromstrand_key</span><span class="p">,</span> <span class="n">annotions_list</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mask_indexes</span><span class="p">[</span><span class="n">chromstrand_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">FeatureIndex</span><span class="p">(</span><span class="n">annotions_list</span><span class="p">)</span>  <span class="c1"># This suould be sorted</span>

        <span class="c1"># Before counting, report how many features where validated</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Summarizing the results of intron validation.&quot;</span><span class="p">)</span>
        <span class="n">n_is_intron</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">n_is_intron_valid</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">unique_valid</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">chromstrand_key</span><span class="p">,</span> <span class="n">feature_index</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_indexes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">ivl</span> <span class="ow">in</span> <span class="n">feature_index</span><span class="o">.</span><span class="n">ivls</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ivl</span><span class="o">.</span><span class="n">kind</span> <span class="o">==</span> <span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;i&quot;</span><span class="p">):</span>
                    <span class="n">n_is_intron</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">ivl</span><span class="o">.</span><span class="n">is_validated</span><span class="p">:</span>
                    <span class="n">n_is_intron_valid</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">unique_valid</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">ivl</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">ivl</span><span class="o">.</span><span class="n">end</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Validated </span><span class="si">{n_is_intron_valid}</span><span class="s2"> introns (of which unique intervals {len(unique_valid)}) out of </span><span class="si">{n_is_intron}</span><span class="s2"> total possible introns (considering each possible transcript models).&quot;</span><span class="p">)</span>

        <span class="n">cell_bcs_order</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">list_spliced_arrays</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">list_unspliced_arrays</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">list_ambiguous_arrays</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">nth</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Loop through the aligment of the samfile</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignments</span><span class="p">(</span><span class="n">samfile</span><span class="p">,</span> <span class="n">unique</span><span class="o">=</span><span class="ow">not</span> <span class="n">multimap</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">)</span> <span class="o">==</span> <span class="n">cell_batch_size</span> <span class="ow">and</span> <span class="n">r</span><span class="o">.</span><span class="n">bc</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">:</span>
                <span class="c1"># Perfrom the molecule counting</span>
                <span class="n">nth</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Counting for batch </span><span class="si">{nth}</span><span class="s2">, containing {len(self.cell_batch)} cells and {len(self.reads_to_count)} reads&quot;</span><span class="p">)</span>
                <span class="n">spliced</span><span class="p">,</span> <span class="n">unspliced</span><span class="p">,</span> <span class="n">ambiguous</span><span class="p">,</span> <span class="n">list_bcs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_cell_batch</span><span class="p">(</span><span class="n">molecules_report</span> <span class="ow">and</span> <span class="p">(</span><span class="n">nth</span> <span class="o">%</span> <span class="mi">10</span> <span class="o">==</span> <span class="mi">0</span><span class="p">))</span>
                <span class="n">cell_bcs_order</span> <span class="o">+=</span> <span class="n">list_bcs</span>
                <span class="n">list_spliced_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spliced</span><span class="p">)</span>
                <span class="n">list_unspliced_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">unspliced</span><span class="p">)</span>
                <span class="n">list_ambiguous_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ambiguous</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="c1"># Drop the counted reads (If there are no other reference to it) and reset the indexes to 0</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">chromstrand_key</span><span class="p">,</span> <span class="n">annotions_ordered_dict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations_by_chrm_strand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">feature_indexes</span><span class="p">[</span><span class="n">chromstrand_key</span><span class="p">]</span><span class="o">.</span><span class="n">reset</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">chromstrand_key</span><span class="p">,</span> <span class="n">annotions_list</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mask_indexes</span><span class="p">[</span><span class="n">chromstrand_key</span><span class="p">]</span><span class="o">.</span><span class="n">reset</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">bc</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Counting molecule for last batch of {len(self.cell_batch)}&quot;</span><span class="p">)</span>
        <span class="n">spliced</span><span class="p">,</span> <span class="n">unspliced</span><span class="p">,</span> <span class="n">ambiguous</span><span class="p">,</span> <span class="n">list_bcs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_cell_batch</span><span class="p">()</span>
        <span class="n">cell_bcs_order</span> <span class="o">+=</span> <span class="n">list_bcs</span>
        <span class="n">list_spliced_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spliced</span><span class="p">)</span>
        <span class="n">list_unspliced_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">unspliced</span><span class="p">)</span>
        <span class="n">list_ambiguous_arrays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ambiguous</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Counting done!&quot;</span><span class="p">)</span></div>
        <span class="k">return</span> <span class="n">list_spliced_arrays</span><span class="p">,</span> <span class="n">list_unspliced_arrays</span><span class="p">,</span> <span class="n">list_ambiguous_arrays</span><span class="p">,</span> <span class="n">cell_bcs_order</span>

<div class="viewcode-block" id="ExInCounter.count_cell_batch"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.count_cell_batch">[docs]</a>    <span class="k">def</span> <span class="nf">count_cell_batch</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecules_report</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;It performs molecule counting for the current batch of cells</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        spliced: np.ndarray</span>
<span class="sd">        unspliced: np.ndarray</span>
<span class="sd">        ambiguous: np.ndarray</span>
<span class="sd">        idx2bc: List[str]</span>
<span class="sd">            list of barcodes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">molitems</span><span class="p">:</span> <span class="n">DefaultDict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Molitem</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">vcy</span><span class="o">.</span><span class="n">Molitem</span><span class="p">)</span>
        <span class="c1"># Sort similarly to what the sort linux command would do. (implemented using Read.__lt__)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="c1"># NOTE: I could start by sorting the reads by chromosome, strand, position but for now let&#39;s see if it is fast without doing do</span>

        <span class="n">repeats_reads_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span><span class="p">:</span>
            <span class="c1"># Consider the correct strand</span>
            <span class="n">ii</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_indexes</span><span class="p">[</span><span class="n">r</span><span class="o">.</span><span class="n">chrom</span> <span class="o">+</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span><span class="p">]</span>
            <span class="n">iim</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_indexes</span><span class="p">[</span><span class="n">r</span><span class="o">.</span><span class="n">chrom</span> <span class="o">+</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span><span class="p">]</span>

            <span class="c1"># Check if read is fully inside a repeat, in that case skip it</span>
            <span class="k">if</span> <span class="n">iim</span><span class="o">.</span><span class="n">has_ivls_enclosing</span><span class="p">(</span><span class="n">r</span><span class="p">):</span>
                <span class="n">repeats_reads_count</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># VERBOSE</span>
                <span class="k">continue</span>

            <span class="c1"># Look for overlap between the intervals and the read</span>
            <span class="n">mappings_record</span> <span class="o">=</span> <span class="n">ii</span><span class="o">.</span><span class="n">find_overlapping_ivls</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mappings_record</span><span class="p">):</span>
                <span class="c1"># logging.debug(&quot;IN: Non empty mapping record&quot;)</span>
                <span class="n">bcumi</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">bc</span> <span class="o">+</span> <span class="n">r</span><span class="o">.</span><span class="n">umi</span>
                <span class="n">molitems</span><span class="p">[</span><span class="n">bcumi</span><span class="p">]</span><span class="o">.</span><span class="n">add_mappings_record</span><span class="p">(</span><span class="n">mappings_record</span><span class="p">)</span>
                <span class="c1"># if len(molitems[bcumi].mappings_record):</span>
                <span class="c1">#     logging.debug(&quot;OUT: Non empty&quot;)</span>
                <span class="c1"># else:</span>
                <span class="c1">#     logging.debug(&quot;OUT: Empty&quot;)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{repeats_reads_count}</span><span class="s2"> reads not considered because fully enclosed in repeat masked regions&quot;</span><span class="p">)</span>  <span class="c1"># VERBOSE</span>
        <span class="c1"># initialize np.ndarray</span>
        <span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">))</span>
        <span class="n">spliced</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">vcy</span><span class="o">.</span><span class="n">LOOM_NUMERIC_DTYPE</span><span class="p">)</span>
        <span class="n">unspliced</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">vcy</span><span class="o">.</span><span class="n">LOOM_NUMERIC_DTYPE</span><span class="p">)</span>
        <span class="n">ambiguous</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">vcy</span><span class="o">.</span><span class="n">LOOM_NUMERIC_DTYPE</span><span class="p">)</span>
        <span class="n">bc2idx</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cell_batch</span><span class="p">))))</span>

        <span class="c1"># After the whole file has been read, do the actual counting</span>
        <span class="k">for</span> <span class="n">bcumi</span><span class="p">,</span> <span class="n">molitem</span> <span class="ow">in</span> <span class="n">molitems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">bc</span> <span class="o">=</span> <span class="n">bcumi</span><span class="p">[:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span><span class="p">]</span>  <span class="c1"># extract the bc part from the bc+umi</span>
            <span class="n">bcidx</span> <span class="o">=</span> <span class="n">bc2idx</span><span class="p">[</span><span class="n">bc</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logic</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">molitem</span><span class="p">,</span> <span class="n">bcidx</span><span class="p">,</span> <span class="n">spliced</span><span class="p">,</span> <span class="n">unspliced</span><span class="p">,</span> <span class="n">ambiguous</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneid2ix</span><span class="p">)</span>
            <span class="c1"># NOTE I need to generalize this to any set of layers</span>
            <span class="c1"># before it was molitem.count(bcidx, spliced, unspliced, ambiguous, self.geneid2ix)</span>

        <span class="k">if</span> <span class="n">molecules_report</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">pickle</span>
            <span class="n">first_cell_batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">molitems</span><span class="o">.</span><span class="n">keys</span><span class="p">()))[:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s2">&quot;pickle_dump&quot;</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;pickle_dump&quot;</span><span class="p">)</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">molitems</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;pickle_dump/molitems_dump_</span><span class="si">{first_cell_batch}</span><span class="s2">.pickle&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads_to_count</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;pickle_dump/reads_to_count</span><span class="si">{first_cell_batch}</span><span class="s2">.pickle&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>

        <span class="n">idx2bc</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="p">:</span> <span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">bc2idx</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        </div>
        <span class="k">return</span> <span class="n">spliced</span><span class="p">,</span> <span class="n">unspliced</span><span class="p">,</span> <span class="n">ambiguous</span><span class="p">,</span> <span class="p">[</span><span class="n">idx2bc</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">idx2bc</span><span class="p">))]</span>

<div class="viewcode-block" id="ExInCounter.pcount"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.pcount">[docs]</a>    <span class="k">def</span> <span class="nf">pcount</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cell_batch_size</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">molecules_report</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_processes</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot; Do the counting of molecules in parallel using multiprocessing</span>
<span class="sd">        &quot;&quot;&quot;</span>
</div>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Implement this using multiprocessiong&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="ExInCounter.pcount_cell_batch"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.pcount_cell_batch">[docs]</a>    <span class="k">def</span> <span class="nf">pcount_cell_batch</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;It performs molecule counting for the current batch of cells</span>
<span class="sd">        &quot;&quot;&quot;</span></div></div>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;This will be a used by .pcount&quot;</span><span class="p">)</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Linnarsson Lab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.13.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>